File: native-views/react-native-segment-slider/ios/HybridSegmentSlider.swift
import Foundation
import UIKit

/// Native segmented slider. Renders the track, fill bar, segment marks, thumb
/// and value bubble as CALayers and handles the drag gesture itself, so a drag
/// never crosses the JS bridge per frame. Mirrors the web variant
/// (`composite/SegmentSlider/index.tsx`) pixel-for-pixel.
final class HybridSegmentSlider: HybridSegmentSliderSpec {

  // MARK: - Geometry (points), identical to the web constants
  private let thumbSize: CGFloat = 16
  private let markSize: CGFloat = 10
  private let haloRing: CGFloat = 4 // translucent halo width beyond each mark
  private var inset: CGFloat { thumbSize / 2 }

  // MARK: - HybridView
  var view: UIView = SegmentSliderUIView()

  // MARK: - Layers
  private let trackLayer = CALayer()
  private let fillLayer = CALayer()
  private let thumbLayer = CALayer()
  private var markLayers: [CALayer] = []
  private let bubbleLayer = CALayer()
  private let bubbleTextLayer = CATextLayer()
  // Translucent highlight shown around the segment node the thumb is over while
  // dragging (mobile analogue of the web mark hover halo). One shared layer,
  // repositioned to the active node; below the marks so the node dot stays crisp.
  private let haloLayer = CALayer()

  // MARK: - Cached colors
  private var cFill = UIColor.clear.cgColor
  private var cTrack = UIColor.clear.cgColor
  private var cThumb = UIColor.white.cgColor
  private var cThumbBorder = UIColor.gray.cgColor
  private var cMarkActive = UIColor.clear.cgColor
  private var cMarkInactive = UIColor.white.cgColor
  private var cMarkBorder = UIColor.gray.cgColor
  private var cBubble = UIColor.black.cgColor
  private var cBubbleText = UIColor.white.cgColor

  // MARK: - Interaction state
  private var currentValue: Double = 0
  private var lastEmitted: Double = .nan
  private var isDragging = false
  private var hasLaidOut = false
  // Tap-vs-drag tracking for `snapTapToSegment`. A gesture starts as a tap
  // candidate; once the finger travels past `tapMoveThreshold` it becomes a
  // drag and the release no longer snaps.
  private var touchStartX: CGFloat = 0
  private var lastTouchX: CGFloat = 0
  private var touchStartTime: CFTimeInterval = 0
  private var hasDragged = false
  private let tapMoveThreshold: CGFloat = 8
  // A tap only snaps when it's quick. Holding longer (without dragging) is a
  // long-press and does nothing — matching the competitor's tap window.
  private let tapMaxDuration: CFTimeInterval = 0.3
  // Segment node the thumb currently sits on (-1 = none). Drives the halo +
  // a light haptic tick fired once each time the thumb enters a new node.
  private var hoverMarkIndex = -1
  private lazy var haptic = UIImpactFeedbackGenerator(style: .light)

  // MARK: - Props (required props are non-optional in the generated spec)
  var value: Double = 0 {
    didSet {
      // While dragging, the finger owns the value — ignore controlled updates
      // (web skips its useLayoutEffect when draggingRef.current is true).
      if !isDragging {
        currentValue = value
        lastEmitted = value
        applyVisual(value)
      }
    }
  }
  var min: Double = 0 { didSet { scheduleLayout() } }
  var max: Double = 100 { didSet { scheduleLayout() } }
  var segments: Double = 0 { didSet { scheduleLayout() } }
  var sliderHeight: Double = 4 { didSet { scheduleLayout() } }
  var disabled: Bool = false {
    didSet {
      view.alpha = disabled ? 0.5 : 1
      view.isUserInteractionEnabled = !disabled
    }
  }
  var showBubble: Bool = true { didSet { scheduleLayout() } }
  var centerOrigin: Bool = false { didSet { scheduleLayout() } }
  // Gates tap snapping only; no visual change on its own, so no didSet.
  var snapTapToSegment: Bool = false
  var epoch: Double = 0 { didSet { scheduleLayout() } }

  var fillColor: String = "" {
    didSet { cFill = parse(fillColor); fillLayer.backgroundColor = cFill; scheduleLayout() }
  }
  var trackColor: String = "" {
    didSet { cTrack = parse(trackColor); trackLayer.backgroundColor = cTrack }
  }
  var thumbColor: String = "" {
    didSet { cThumb = parse(thumbColor); thumbLayer.backgroundColor = cThumb }
  }
  var thumbBorderColor: String = "" {
    didSet { cThumbBorder = parse(thumbBorderColor); thumbLayer.borderColor = cThumbBorder }
  }
  var markActiveColor: String = "" {
    didSet { cMarkActive = parse(markActiveColor); scheduleLayout() }
  }
  var markInactiveColor: String = "" {
    didSet { cMarkInactive = parse(markInactiveColor); scheduleLayout() }
  }
  var markBorderColor: String = "" {
    didSet { cMarkBorder = parse(markBorderColor); scheduleLayout() }
  }
  var bubbleColor: String = "" {
    didSet { cBubble = parse(bubbleColor); bubbleLayer.backgroundColor = cBubble }
  }
  var bubbleTextColor: String = "" {
    didSet { cBubbleText = parse(bubbleTextColor); bubbleTextLayer.foregroundColor = cBubbleText }
  }

  // MARK: - Callbacks
  var onChange: ((Double) -> Void)?
  var onSlideStart: (() -> Void)?
  var onSlideComplete: (() -> Void)?

  // MARK: - Init
  override init() {
    super.init()
    view.backgroundColor = .clear
    view.isUserInteractionEnabled = true

    trackLayer.cornerRadius = CGFloat(sliderHeight) / 2
    fillLayer.cornerRadius = CGFloat(sliderHeight) / 2

    thumbLayer.cornerRadius = thumbSize / 2
    thumbLayer.borderWidth = 1
    thumbLayer.shadowColor = UIColor.black.cgColor
    thumbLayer.shadowOpacity = 0.08
    thumbLayer.shadowRadius = 1
    thumbLayer.shadowOffset = CGSize(width: 0, height: 1)

    bubbleLayer.cornerRadius = 4
    bubbleLayer.opacity = 0
    bubbleTextLayer.contentsScale = UIScreen.main.scale
    bubbleTextLayer.alignmentMode = .center
    bubbleTextLayer.fontSize = 12
    bubbleTextLayer.font = UIFont.systemFont(ofSize: 12, weight: .semibold)
    bubbleLayer.addSublayer(bubbleTextLayer)

    let haloSize = markSize + haloRing * 2
    haloLayer.bounds = CGRect(x: 0, y: 0, width: haloSize, height: haloSize)
    haloLayer.cornerRadius = haloSize / 2
    haloLayer.opacity = 0

    view.layer.addSublayer(trackLayer)
    view.layer.addSublayer(fillLayer)
    view.layer.addSublayer(thumbLayer)
    view.layer.addSublayer(haloLayer)
    view.layer.addSublayer(bubbleLayer)

    if let v = view as? SegmentSliderUIView {
      v.onLayout = { [weak self] in self?.performLayout() }
      v.onTouchBegan = { [weak self] x in self?.handleTouchBegan(x) }
      v.onTouchMoved = { [weak self] x in self?.handleTouchMoved(x) }
      v.onTouchEnded = { [weak self] in self?.handleTouchEnded() }
    }
  }

  // MARK: - Color helper
  private func parse(_ s: String) -> CGColor {
    SegmentSliderColorParser.parse(s).cgColor
  }

  private func scheduleLayout() {
    view.setNeedsLayout()
  }

  // MARK: - Geometry helpers
  private var trackWidth: CGFloat { Swift.max(view.bounds.width - thumbSize, 0) }
  private var trackLeft: CGFloat { inset }
  private var range: Double { let r = max - min; return r == 0 ? 1 : r }
  private var markCount: Int { segments >= 1 ? Int(segments) + 1 : 0 }

  /// value -> fraction 0..1
  private func valueToFrac(_ v: Double) -> CGFloat {
    let clamped = Swift.max(min, Swift.min(max, v))
    return CGFloat((clamped - min) / range)
  }

  private func valueFromX(_ x: CGFloat) -> Double {
    let w = trackWidth
    if w <= 0 { return currentValue }
    let clampedX = Swift.max(trackLeft, Swift.min(trackLeft + w, x))
    let frac = Double((clampedX - trackLeft) / w)
    return min + frac * (max - min)
  }

  // MARK: - Layout
  private func performLayout() {
    let bounds = view.bounds
    guard bounds.width > 0 else { return }

    syncMarkLayers(markCount)

    let h = CGFloat(sliderHeight)
    let trackTop = bounds.midY - h / 2

    CATransaction.begin()
    CATransaction.setDisableActions(true)

    trackLayer.frame = CGRect(x: trackLeft, y: trackTop, width: trackWidth, height: h)
    trackLayer.cornerRadius = h / 2
    trackLayer.backgroundColor = cTrack
    fillLayer.cornerRadius = h / 2

    // Marks sit on the track line, centred vertically on it.
    for i in 0..<markLayers.count {
      let layer = markLayers[i]
      let frac = markCount > 1 ? CGFloat(i) / CGFloat(markCount - 1) : 0
      let cx = trackLeft + frac * trackWidth
      layer.bounds = CGRect(x: 0, y: 0, width: markSize, height: markSize)
      layer.position = CGPoint(x: cx, y: bounds.midY)
      layer.cornerRadius = markSize / 2
      layer.borderWidth = 1
    }

    // Thumb sizing (position set in applyVisual).
    thumbLayer.bounds = CGRect(x: 0, y: 0, width: thumbSize, height: thumbSize)
    thumbLayer.cornerRadius = thumbSize / 2
    thumbLayer.backgroundColor = cThumb
    thumbLayer.borderColor = cThumbBorder

    CATransaction.commit()

    hasLaidOut = true
    applyVisual(currentValue)
  }

  private func syncMarkLayers(_ count: Int) {
    if markLayers.count < count {
      for _ in markLayers.count..<count {
        let l = CALayer()
        l.borderWidth = 1
        // Marks render ABOVE the thumb + halo (web parity: mark zIndex 3 >
        // thumb 2). So when the thumb sits on a segment node the node's dot
        // shows through the thumb's centre — the "white ring + dark centre"
        // bullseye look. Inserting above the halo keeps order
        // track < fill < thumb < halo < marks < bubble.
        view.layer.insertSublayer(l, above: haloLayer)
        markLayers.append(l)
      }
    } else if markLayers.count > count {
      for l in markLayers[count...] { l.removeFromSuperlayer() }
      markLayers.removeLast(markLayers.count - count)
    }
  }

  // MARK: - Visual application (no implicit animations)
  private func applyVisual(_ v: Double) {
    guard hasLaidOut, view.bounds.width > 0 else { return }
    let bounds = view.bounds
    let h = CGFloat(sliderHeight)
    let trackTop = bounds.midY - h / 2
    let frac = valueToFrac(v)
    let thumbX = trackLeft + frac * trackWidth

    CATransaction.begin()
    CATransaction.setDisableActions(true)

    // Fill bar.
    if centerOrigin {
      if v == 0 {
        fillLayer.frame = CGRect(x: trackLeft, y: trackTop, width: 0, height: h)
      } else {
        let centerFrac = valueToFrac(0)
        let leftFrac = Swift.min(frac, centerFrac)
        let widthFrac = abs(frac - centerFrac)
        fillLayer.frame = CGRect(
          x: trackLeft + leftFrac * trackWidth, y: trackTop,
          width: widthFrac * trackWidth, height: h)
      }
    } else {
      fillLayer.frame = CGRect(
        x: trackLeft, y: trackTop, width: frac * trackWidth, height: h)
    }
    fillLayer.backgroundColor = cFill

    // Marks active states.
    let centerFrac = valueToFrac(0)
    for i in 0..<markLayers.count {
      let markFrac = markLayers.count > 1 ? CGFloat(i) / CGFloat(markLayers.count - 1) : 0
      var active: Bool
      if centerOrigin {
        if v == 0 {
          active = false
        } else if v > 0 {
          active = markFrac > centerFrac && markFrac <= frac
        } else {
          active = markFrac >= frac && markFrac < centerFrac
        }
      } else {
        active = markFrac <= frac
      }
      let layer = markLayers[i]
      layer.backgroundColor = active ? cMarkActive : cMarkInactive
      layer.borderColor = active ? cMarkActive : cMarkBorder
    }

    // Thumb position (keep current scale transform).
    thumbLayer.position = CGPoint(x: thumbX, y: bounds.midY)

    // Bubble.
    if showBubble {
      let text = "\(Int(v.rounded()))%"
      bubbleTextLayer.string = text
      let font = UIFont.systemFont(ofSize: 12, weight: .semibold)
      let textW = (text as NSString)
        .size(withAttributes: [.font: font]).width
      let padH: CGFloat = 8
      let bw = ceil(textW) + padH * 2
      let bh: CGFloat = 20
      let by = trackTop - 8 - bh
      bubbleLayer.frame = CGRect(x: thumbX - bw / 2, y: by, width: bw, height: bh)
      bubbleTextLayer.frame = CGRect(x: 0, y: 2, width: bw, height: 16)
    }

    CATransaction.commit()

    updateHover(thumbX: thumbX)
  }

  // MARK: - Node halo + haptic (mobile analogue of the web mark hover)
  /// Highlights the node the thumb is over and fires a light haptic tick once
  /// per node entered. Only while dragging — controlled value syncs just keep
  /// the halo hidden. Does NOT snap the value (parity with web/perp).
  private func updateHover(thumbX: CGFloat) {
    guard isDragging, markCount > 0 else {
      if hoverMarkIndex != -1 { hoverMarkIndex = -1; setHalo(visible: false, index: 0) }
      return
    }
    let threshold: CGFloat = 12 // half of the 24pt mark hit area
    var best = -1
    var bestDist = threshold
    for i in 0..<markCount {
      let frac = markCount > 1 ? CGFloat(i) / CGFloat(markCount - 1) : 0
      let mx = trackLeft + frac * trackWidth
      let d = abs(thumbX - mx)
      if d <= bestDist { bestDist = d; best = i }
    }
    guard best != hoverMarkIndex else { return }
    hoverMarkIndex = best
    if best == -1 {
      setHalo(visible: false, index: 0)
    } else {
      setHalo(visible: true, index: best)
      haptic.impactOccurred(intensity: 0.6)
      haptic.prepare()
    }
  }

  private func setHalo(visible: Bool, index: Int) {
    CATransaction.begin()
    CATransaction.setDisableActions(true)
    if visible {
      let frac = markCount > 1 ? CGFloat(index) / CGFloat(markCount - 1) : 0
      let mx = trackLeft + frac * trackWidth
      haloLayer.position = CGPoint(x: mx, y: view.bounds.midY)
      haloLayer.backgroundColor = UIColor(cgColor: cMarkActive)
        .withAlphaComponent(0.25).cgColor
      haloLayer.opacity = 1
    } else {
      haloLayer.opacity = 0
    }
    CATransaction.commit()
  }

  // MARK: - Touch handling
  /// True when tap-snapping should defer the value commit: enabled and there
  /// are marks to snap to.
  private var snapTapsActive: Bool { snapTapToSegment && markCount > 1 }

  private func handleTouchBegan(_ x: CGFloat) {
    guard !disabled else { return }
    isDragging = true
    hasDragged = false
    touchStartX = x
    lastTouchX = x
    touchStartTime = CACurrentMediaTime()
    haptic.prepare()
    onSlideStart?()
    animateThumbScale(1.15)
    setBubbleVisible(true)
    // Tap-snap mode: defer any value change until we know this is a tap or a
    // drag, so a tap never jumps to the raw touch point first.
    if !snapTapsActive {
      setValue(valueFromX(x))
    }
  }

  private func handleTouchMoved(_ x: CGFloat) {
    guard isDragging else { return }
    lastTouchX = x
    if !hasDragged, abs(x - touchStartX) > tapMoveThreshold {
      hasDragged = true
    }
    // Still a tap candidate in snap mode → keep the thumb put.
    if snapTapsActive, !hasDragged { return }
    setValue(valueFromX(x))
  }

  private func handleTouchEnded() {
    guard isDragging else { return }
    isDragging = false
    // A quick tap (no drag) lands directly on the nearest segment node in a
    // single move — no detour through the raw touch point. A long press (held
    // past tapMaxDuration without dragging) does nothing. A drag keeps its
    // free value.
    if snapTapsActive, !hasDragged {
      let pressDuration = CACurrentMediaTime() - touchStartTime
      if pressDuration <= tapMaxDuration {
        setValue(snapValueToNearestMark(valueFromX(lastTouchX)))
      }
    }
    hoverMarkIndex = -1
    setHalo(visible: false, index: 0)
    animateThumbScale(1.0)
    setBubbleVisible(false)
    onSlideComplete?()
  }

  /// Snaps a value to the value of the nearest segment mark. Caller guarantees
  /// `markCount > 1`.
  private func snapValueToNearestMark(_ v: Double) -> Double {
    let frac = Double(valueToFrac(v))
    let idx = (frac * Double(markCount - 1)).rounded()
    let snappedFrac = idx / Double(markCount - 1)
    return min + snappedFrac * (max - min)
  }

  private func setValue(_ raw: Double) {
    let clamped = Swift.max(min, Swift.min(max, raw))
    let rounded = clamped.rounded()
    currentValue = rounded
    applyVisual(rounded)
    if rounded != lastEmitted {
      lastEmitted = rounded
      onChange?(rounded)
    }
  }

  private func animateThumbScale(_ s: CGFloat) {
    CATransaction.begin()
    CATransaction.setAnimationDuration(0.12)
    thumbLayer.transform = CATransform3DMakeScale(s, s, 1)
    CATransaction.commit()
  }

  private func setBubbleVisible(_ visible: Bool) {
    guard showBubble else { return }
    CATransaction.begin()
    CATransaction.setAnimationDuration(0.15)
    bubbleLayer.opacity = visible ? 1 : 0
    CATransaction.commit()
  }
}

File: native-views/react-native-segment-slider/android/src/main/java/com/margelo/nitro/segmentslider/HybridSegmentSlider.kt
package com.margelo.nitro.segmentslider

import android.animation.ValueAnimator
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.view.HapticFeedbackConstants
import android.view.MotionEvent
import android.view.View
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext

/**
 * Native segmented slider. Draws the track, fill bar, segment marks, thumb and
 * value bubble in [View.onDraw] and handles the drag in [View.onTouchEvent], so
 * a drag never crosses the JS bridge per frame. Mirrors the web variant
 * (`composite/SegmentSlider/index.tsx`) pixel-for-pixel.
 */
@DoNotStrip
class HybridSegmentSlider(val context: ThemedReactContext) : HybridSegmentSliderSpec() {

  private val density = context.resources.displayMetrics.density

  // Geometry (dp), identical to the web constants.
  private val thumbSizePx = 16f * density
  private val markSizePx = 10f * density
  private val insetPx = thumbSizePx / 2f
  private val haloRingPx = 4f * density // translucent halo width beyond each mark

  // Paints.
  private val trackPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }
  private val fillPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }
  private val thumbFillPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }
  private val thumbStrokePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    style = Paint.Style.STROKE; strokeWidth = density
  }
  private val markFillPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }
  private val markStrokePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    style = Paint.Style.STROKE; strokeWidth = density
  }
  // Translucent highlight around the node the thumb is over while dragging.
  private val markHaloPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }
  private val bubblePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { style = Paint.Style.FILL }
  private val bubbleTextPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    textAlign = Paint.Align.CENTER
    textSize = 12f * density
    isFakeBoldText = true
  }
  private val rect = RectF()

  // Cached colors.
  private var cFill = 0
  private var cTrack = 0
  private var cThumb = 0
  private var cThumbBorder = 0
  private var cMarkActive = 0
  private var cMarkInactive = 0
  private var cMarkBorder = 0

  // Interaction state.
  private var currentValue = 0.0
  private var lastEmitted = Double.NaN
  private var isDragging = false
  private var thumbScale = 1f
  private var thumbScaleAnimator: ValueAnimator? = null
  private var isDisposed = false
  // Segment node the thumb currently sits on (-1 = none). Drives the halo +
  // a light haptic tick fired once each time the thumb enters a new node.
  private var hoverMarkIndex = -1
  // Tap-vs-drag tracking for `snapTapToSegment`. A gesture starts as a tap
  // candidate; once the finger travels past `tapMoveThresholdPx` it becomes a
  // drag and the release no longer snaps.
  private var touchStartX = 0f
  private var lastTouchX = 0f
  private var hasDragged = false
  private val tapMoveThresholdPx = 8f * density
  // A tap only snaps when it's quick. Holding longer (without dragging) is a
  // long-press and does nothing — matching the competitor's tap window.
  private val tapMaxDurationMs = 300L

  override val view: View = object : View(context) {
    override fun onDraw(canvas: Canvas) {
      drawSlider(canvas)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
      if (disabled || isDisposed) return false
      // Use actionMasked so multi-touch pointer events (ACTION_POINTER_UP /
      // pointer-index > 0) still resolve to the right action and never strand
      // the drag state or the disallow-intercept flag.
      when (event.actionMasked) {
        MotionEvent.ACTION_DOWN -> {
          isDragging = true
          hasDragged = false
          touchStartX = event.x
          lastTouchX = event.x
          onSlideStart?.invoke()
          animateThumbScale(1.15f)
          // Tap-snap mode: defer the value commit until we know this is a tap or
          // a drag, so a tap never jumps to the raw touch point first.
          if (!snapTapsActive) commitDragValue(valueFromX(event.x))
          parent?.requestDisallowInterceptTouchEvent(true)
          return true
        }
        MotionEvent.ACTION_MOVE -> {
          if (isDragging) {
            lastTouchX = event.x
            if (!hasDragged && Math.abs(event.x - touchStartX) > tapMoveThresholdPx) {
              hasDragged = true
            }
            // Still a tap candidate in snap mode → keep the thumb put.
            if (!(snapTapsActive && !hasDragged)) commitDragValue(valueFromX(event.x))
          }
          return true
        }
        MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
          if (isDragging) {
            isDragging = false
            // A quick tap (no drag) lands directly on the nearest segment node
            // in a single move — no detour through the raw touch point. A long
            // press (held past tapMaxDurationMs without dragging) does nothing.
            // A drag keeps its free value.
            if (snapTapsActive && !hasDragged &&
              (event.eventTime - event.downTime) <= tapMaxDurationMs
            ) {
              commitDragValue(snapValueToNearestMark(valueFromX(lastTouchX)))
            }
            hoverMarkIndex = -1
            animateThumbScale(1f)
            view.invalidate()
            onSlideComplete?.invoke()
          }
          parent?.requestDisallowInterceptTouchEvent(false)
          return true
        }
      }
      return super.onTouchEvent(event)
    }
  }.apply {
    isClickable = true
    isFocusable = false
    isHapticFeedbackEnabled = true
  }

  // MARK: - Props
  // Guards the apply-once semantics of [defaultValue]: only the first prop
  // commit seeds the initial value. After that the view is uncontrolled and JS
  // drives it via the imperative [setValue] method.
  private var didApplyDefault = false

  override var defaultValue: Double = 0.0
    get() = field
    set(v) {
      field = v
      if (!didApplyDefault) {
        didApplyDefault = true
        syncToValue(v)
      }
    }

  override var min: Double = 0.0
    get() = field
    set(v) { field = v; invalidateIfAlive() }

  override var max: Double = 100.0
    get() = field
    set(v) { field = v; invalidateIfAlive() }

  override var segments: Double = 0.0
    get() = field
    set(v) { field = v; invalidateIfAlive() }

  override var sliderHeight: Double = 4.0
    get() = field
    set(v) { field = v; invalidateIfAlive() }

  override var disabled: Boolean = false
    get() = field
    set(v) {
      field = v
      view.alpha = if (v) 0.5f else 1f
      invalidateIfAlive()
    }

  override var showBubble: Boolean = true
    get() = field
    set(v) { field = v; invalidateIfAlive() }

  override var centerOrigin: Boolean = false
    get() = field
    set(v) { field = v; invalidateIfAlive() }

  // Gates tap snapping only; no visual change on its own.
  override var snapTapToSegment: Boolean = false
    get() = field
    set(v) { field = v }

  override var alignSegmentMarksToIntegerValues: Boolean = false
    get() = field
    set(v) { field = v; invalidateIfAlive() }

  override var fillColor: String = ""
    get() = field
    set(v) { field = v; cFill = parse(v); fillPaint.color = cFill; invalidateIfAlive() }

  override var trackColor: String = ""
    get() = field
    set(v) { field = v; cTrack = parse(v); trackPaint.color = cTrack; invalidateIfAlive() }

  override var thumbColor: String = ""
    get() = field
    set(v) { field = v; cThumb = parse(v); thumbFillPaint.color = cThumb; invalidateIfAlive() }

  override var thumbBorderColor: String = ""
    get() = field
    set(v) { field = v; cThumbBorder = parse(v); thumbStrokePaint.color = cThumbBorder; invalidateIfAlive() }

  override var markActiveColor: String = ""
    get() = field
    set(v) { field = v; cMarkActive = parse(v); invalidateIfAlive() }

  override var markInactiveColor: String = ""
    get() = field
    set(v) { field = v; cMarkInactive = parse(v); invalidateIfAlive() }

  override var markBorderColor: String = ""
    get() = field
    set(v) { field = v; cMarkBorder = parse(v); invalidateIfAlive() }

  override var bubbleColor: String = ""
    get() = field
    set(v) { field = v; bubblePaint.color = parse(v); invalidateIfAlive() }

  override var bubbleTextColor: String = ""
    get() = field
    set(v) { field = v; bubbleTextPaint.color = parse(v); invalidateIfAlive() }

  // MARK: - Callbacks
  override var onChange: ((value: Double) -> Unit)? = null
  override var onSlideStart: (() -> Unit)? = null
  override var onSlideComplete: (() -> Unit)? = null

  // MARK: - Imperative methods
  // Set the value from JS without a Fabric prop commit. Called on the JS
  // thread, so hop to the UI thread before touching the view / drawing state
  // (mirrors react-native-perp-depth-bar's imperative methods). Does NOT fire
  // onChange — syncToValue updates lastEmitted so the caller's own value never
  // echoes back.
  override fun setValue(value: Double) {
    view.post {
      if (!isDisposed) syncToValue(value)
    }
  }

  private fun parse(v: String) = SegmentSliderColorParser.parse(v)

  private fun invalidateIfAlive() {
    if (!isDisposed) view.invalidate()
  }

  // MARK: - Geometry helpers
  private val trackWidth: Float get() = (view.width - thumbSizePx).coerceAtLeast(0f)
  private val rangeSafe: Double get() = (max - min).let { if (it == 0.0) 1.0 else it }
  private val markCount: Int get() = if (segments >= 1) segments.toInt() + 1 else 0

  private fun segmentValue(index: Int): Double {
    if (markCount <= 1) return min
    val fraction = index.toDouble() / (markCount - 1)
    val value = min + fraction * (max - min)
    return if (alignSegmentMarksToIntegerValues) Math.rint(value) else value
  }

  private fun markFraction(index: Int): Float = valueToFrac(segmentValue(index))

  private fun nearestMarkIndex(value: Double): Int {
    val valueFraction = valueToFrac(value)
    var nearestIndex = 0
    var nearestDistance = Float.POSITIVE_INFINITY
    for (index in 0 until markCount) {
      val distance = Math.abs(valueFraction - markFraction(index))
      if (distance <= nearestDistance) {
        nearestDistance = distance
        nearestIndex = index
      }
    }
    return nearestIndex
  }

  /** True when tap-snapping should defer the value commit: enabled with marks to snap to. */
  private val snapTapsActive: Boolean get() = snapTapToSegment && markCount > 1

  private fun valueToFrac(v: Double): Float {
    val clamped = v.coerceIn(min, max)
    return ((clamped - min) / rangeSafe).toFloat()
  }

  private fun valueFromX(x: Float): Double {
    val w = trackWidth
    if (w <= 0f) return currentValue
    val clampedX = x.coerceIn(insetPx, insetPx + w)
    val frac = ((clampedX - insetPx) / w).toDouble()
    return min + frac * (max - min)
  }

  /** Snaps a value to the value of the nearest segment mark. Caller guarantees `markCount > 1`. */
  private fun snapValueToNearestMark(v: Double): Double {
    return segmentValue(nearestMarkIndex(v))
  }

  /** Hard re-sync: snap the thumb to [v], overriding any in-flight drag. */
  private fun syncToValue(v: Double) {
    if (isDisposed) return
    // An imperative setValue (or the initial defaultValue) wins over the
    // current gesture: drop the drag so a queued ACTION_MOVE can't immediately
    // re-overwrite the thumb, and clear the transient hover/halo so the snapped
    // position renders clean.
    isDragging = false
    hoverMarkIndex = -1
    currentValue = v
    lastEmitted = v
    view.invalidate()
  }

  private fun commitDragValue(raw: Double) {
    val rounded = raw.coerceIn(min, max).let { Math.round(it).toDouble() }
    currentValue = rounded
    updateHover()
    view.invalidate()
    if (rounded != lastEmitted) {
      lastEmitted = rounded
      onChange?.invoke(rounded)
    }
  }

  /// Highlights the node the thumb is over and fires a light haptic tick once
  /// per node entered. Only while dragging; does NOT snap (parity with web/perp).
  private fun updateHover() {
    val tw = trackWidth
    if (!isDragging || markCount <= 0 || tw <= 0f) {
      hoverMarkIndex = -1
      return
    }
    val thumbX = insetPx + valueToFrac(currentValue) * tw
    val threshold = 12f * density // half of the 24dp mark hit area
    var best = -1
    var bestDist = threshold
    for (i in 0 until markCount) {
      val mf = markFraction(i)
      val mx = insetPx + mf * tw
      val d = Math.abs(thumbX - mx)
      if (d <= bestDist) { bestDist = d; best = i }
    }
    if (best != hoverMarkIndex) {
      hoverMarkIndex = best
      if (best != -1) {
        view.performHapticFeedback(HapticFeedbackConstants.CLOCK_TICK)
      }
    }
  }

  private fun animateThumbScale(target: Float) {
    thumbScaleAnimator?.cancel()
    thumbScaleAnimator = ValueAnimator.ofFloat(thumbScale, target).apply {
      duration = 120L
      addUpdateListener { a ->
        if (isDisposed) return@addUpdateListener
        thumbScale = a.animatedValue as Float
        view.invalidate()
      }
      start()
    }
  }

  // MARK: - Drawing
  private fun drawSlider(canvas: Canvas) {
    val w = view.width.toFloat()
    if (w <= 0f) return
    val h = (sliderHeight * density).toFloat()
    val midY = view.height / 2f
    val trackTop = midY - h / 2f
    val tw = trackWidth
    val frac = valueToFrac(currentValue)
    val radius = h / 2f

    // Track.
    rect.set(insetPx, trackTop, insetPx + tw, trackTop + h)
    canvas.drawRoundRect(rect, radius, radius, trackPaint)

    // Fill.
    if (centerOrigin) {
      if (currentValue != 0.0) {
        val centerFrac = valueToFrac(0.0)
        val leftFrac = minOf(frac, centerFrac)
        val widthFrac = Math.abs(frac - centerFrac)
        rect.set(insetPx + leftFrac * tw, trackTop, insetPx + (leftFrac + widthFrac) * tw, trackTop + h)
        canvas.drawRoundRect(rect, radius, radius, fillPaint)
      }
    } else if (frac > 0f) {
      rect.set(insetPx, trackTop, insetPx + frac * tw, trackTop + h)
      canvas.drawRoundRect(rect, radius, radius, fillPaint)
    }

    // Thumb (drawn BELOW the marks so a node under the thumb shows through its
    // centre — the "white ring + dark centre" bullseye, matching web where the
    // mark zIndex 3 sits above the thumb zIndex 2).
    val thumbX = insetPx + frac * tw
    val thumbR = thumbSizePx / 2f * thumbScale
    canvas.drawCircle(thumbX, midY, thumbR, thumbFillPaint)
    canvas.drawCircle(thumbX, midY, thumbR - density / 2f, thumbStrokePaint)

    // Node halo (drag-time highlight) — drawn under the marks so the node dot
    // stays crisp on top. Translucent markActive color (25% alpha).
    if (isDragging && hoverMarkIndex in 0 until markCount) {
      val mf = markFraction(hoverMarkIndex)
      val hx = insetPx + mf * tw
      markHaloPaint.color = (cMarkActive and 0x00FFFFFF) or (0x3F shl 24)
      canvas.drawCircle(hx, midY, markSizePx / 2f + haloRingPx, markHaloPaint)
    }

    // Marks (on top of the thumb).
    val centerFrac = valueToFrac(0.0)
    val markR = markSizePx / 2f
    for (i in 0 until markCount) {
      val markFrac = markFraction(i)
      val cx = insetPx + markFrac * tw
      val active = when {
        centerOrigin && currentValue == 0.0 -> false
        centerOrigin && currentValue > 0 -> markFrac > centerFrac && markFrac <= frac
        centerOrigin && currentValue < 0 -> markFrac >= frac && markFrac < centerFrac
        else -> markFrac <= frac
      }
      markFillPaint.color = if (active) cMarkActive else cMarkInactive
      markStrokePaint.color = if (active) cMarkActive else cMarkBorder
      canvas.drawCircle(cx, midY, markR - density / 2f, markFillPaint)
      canvas.drawCircle(cx, midY, markR - density / 2f, markStrokePaint)
    }

    // Bubble.
    if (isDragging && showBubble) {
      val text = "${Math.round(currentValue)}%"
      val textW = bubbleTextPaint.measureText(text)
      val padH = 8f * density
      val bw = textW + padH * 2
      val bh = 20f * density
      val left = thumbX - bw / 2f
      val top = trackTop - 8f * density - bh
      rect.set(left, top, left + bw, top + bh)
      canvas.drawRoundRect(rect, 4f * density, 4f * density, bubblePaint)
      val baseline = top + bh / 2f - (bubbleTextPaint.descent() + bubbleTextPaint.ascent()) / 2f
      canvas.drawText(text, thumbX, baseline, bubbleTextPaint)
    }
  }

  override fun afterUpdate() {
    super.afterUpdate()
    invalidateIfAlive()
  }

  override fun dispose() {
    isDisposed = true
    thumbScaleAnimator?.cancel()
    thumbScaleAnimator = null
  }
}
